src/lib/y2storage/btrfs_qgroup.rb

```diff
@@ -0,0 +1,115 @@
+# Copyright (c) [2020] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "y2storage/storage_class_wrapper"
+require "y2storage/disk_size"
+
+module Y2Storage
+  # A qgroup of a Btrfs filesystem
+  #
+  # This is a wrapper for Storage::BtrfsQgroup.
+  #
+  # In Btrfs, the qgroups hold the quota information for a given set of subvolumes.
+  # Although they can be organized in a tree hierarchy, we are only considering the
+  # first level (level 0, actually), so they are associated to a single subvolume.
+  #
+  # @see man btrfs-qgroup
+  class BtrfsQgroup < Device
+    wrap_class Storage::BtrfsQgroup
+
+    storage_forward :storage_id, to: :id
+    private :storage_id
+
+    # Id of the group including the level and the qgroup id
+    #
+    # For qgroups of level 0, the second term (the qgroup id) matches the id of
+    # the corresponding subvolume.
+    #
+    # @return [Array<Integer>] includes two terms, level and id
+    def id
+      [storage_id.first, storage_id.second]
+    end
+
+    # @!method referenced
+    #   Size of the referenced space of the qgroup
+    #
+    #   @return [DiskSize] zero if the size is not known (the qgroup was created after probing)
```

Comment:
This means that you can't tell if the qgroup ran out of disk space or if it's just unknown. In the general context of quotas this sounds like an accident waiting to happen. Wouldn't it be better to return `nil` if it's unknown? or -1?

Reply:
In fact, it means you can't distinguish between this two cases:

- the subvolume exists for real (ie. is probed) but is empty
- the subvolume is still not created in the system (so the space it will use is unknown yet)

First of all, that's consistent with what libstorage-ng returns. On the other hand, I don't think it's specially dangerous.

Reply:
OK, if libstorage-ng doesn't make the difference, we can't do much here anyway.